src/lib/statistics/statistics_objects/abstract_histogram.hpp

```diff
@@ -0,0 +1,256 @@
+   * without null values.
+   */
+  virtual HistogramCountType total_count() const = 0;
+
+  /**
+   * Returns the number of distinct values represented in the histogram.
+   * This is equal to the number of distinct values in the segment during creation.
+   */
+  virtual HistogramCountType total_distinct_count() const = 0;
+
+  /**
+   * Returns the smallest value in the bin.
+   */
+  virtual T bin_minimum(const BinID index) const = 0;
+
+  /**
+   * Returns the largest value in a bin.
+   */
+  virtual T bin_maximum(const BinID index) const = 0;
+
+  /**
+   * Returns the number of values in a bin.
+   */
+  virtual HistogramCountType bin_height(const BinID index) const = 0;
+
+  /**
+   * Returns the number of distinct values in a bin.
+   */
+  virtual HistogramCountType bin_distinct_count(const BinID index) const = 0;
+
+  /**
+   * Returns the width of a bin.
+   * This method is specialized for strings to return a numerical width.
+   */
+  virtual HistogramWidthType bin_width(const BinID index) const;
+
+  /**
+   * Helper function
+   * @return {bin_minimum(index), bin_maximum(index), bin_height(index), bin_distinct_count(index)}
+   */
+  HistogramBin<T> bin(const BinID index) const;
+
+  /**
+   * Returns the share of the value range of a bin that are smaller (or equals) than `value`.
+   * This method is specialized for strings.
+   * @{
+   */
+  float bin_ratio_less_than(const BinID bin_id, const T& value) const;
+  float bin_ratio_less_than_equals(const BinID bin_id, const T& value) const;
+  /** @} */
+
+ protected:
+  Cardinality _invert_estimate(const Cardinality& estimate) const;
+  CardinalityAndDistinctCountEstimate _invert_estimate(const CardinalityAndDistinctCountEstimate& estimate) const;
+
+  /**
+   * Returns whether a given predicate type and its parameter(s) can belong to a bin or not.
+   * Specifically, if this method returns true, the predicate does not yield any results.
+   * If this method returns false, the predicate might yield results.
+   * This method is specialized for strings to handle predicates uniquely applicable to string columns.
+   */
+  bool _does_not_contain(const PredicateCondition predicate_condition, const AllTypeVariant& variant_value,
+                         const std::optional<AllTypeVariant>& variant_value2 = std::nullopt) const;
+
+  /**
+   * Returns whether a given predicate type and its parameter(s) can belong to a bin or not
+   * for predicate types supported by all data types.
+   */
+  bool _general_does_not_contain(const PredicateCondition predicate_condition, const AllTypeVariant& variant_value,
+                                 const std::optional<AllTypeVariant>& variant_value2 = std::nullopt) const;
+
+  /**
+   * Returns the id of the bin that holds the given `value`.
+   * Returns INVALID_BIN_ID if `value` does not belong to any bin.
+   */
+  virtual BinID _bin_for_value(const T& value) const = 0;
```

Comment:
legacy. But. no one would use them right now, and if they were public I'd need to add tests for them, which would - since they're only used internally anyway - just be boilerplate.